while-loops/problem-set-while-loops.md

```diff
@@ -0,0 +1,849 @@
+
+##### !end-question
+
+##### !answer
+
+1. while True:
+1.     user_input = input('enter input: ')
+1.     if user_input == 'done':
+1.         break
+1.     elif user_input == 'skip':
+1.         continue
+1.     print(f'you entered {user_input}')
+
+##### !end-answer
+
+##### !explanation
+
+This implementation uses `while True: ...` to run the loop forever. It exits the loop by using the `break` keyword. And it skips the `print` by using the `continue` keyword. There are many other approaches that are just as valid. For an extra challenge, try implementing this loop yourself without using `break` or `continue`!
+
+##### !end-explanation
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+<!-- prettier-ignore-end -->
+
+
+<!-- prettier-ignore-start -->
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: code-snippet
+* language: python3.6
+* id: 85dffbd3-2030-44b4-8b9f-cc86f0c7e9d1
+* title: While Loops
+
+##### !question
+
+Let's implement a function called `sidewinder`. It will take a parameter `number_list` which contains a list of the _next_ index in the list to visit. Starting from position `0`, we will look in the current position, get the value stored there, and use it as the _next_ position to visit. We will continue doing this until we find a value of `0`, which means we have found the end of the index chain. Then we'll return the position where we found the `0`.
+
+For example, if we received `[2, 0, 1]` as our `number_list`, starting from position `0` we find a `2`. We look in position `2` and find a `1`. So we look in position `1` and find `0`. `0` marks the end of the chain, so we return `1`, the position we just looked in.
+
+The diagram below shows a longer example. Notice that there are _multiple_ zeros in the list, so we _cannot_ iterate through in order and return the position of the first zero that we find.
+
+<div style="max-width: 700px; margin: 0 auto -28px auto; text-align: center;">
+<img alt="How to interpret the array input to sidewinder" src="../assets/while-loops_problem-set-while-loops_sidewinder.png">
+</div>
+*Fig. Given the list `[4, 9, 0, 0, 1, 8, 0, 0, 11, 5, 0, 10]` for `number_list`, this diagram shows how to move through the list values to find the 'end.' Starting from index `0`, we find the value `4`. We go to index `4` where we find the value `1`. We go to index `1` where we find the value `9`, and so on. Eventually we reach index `10` where we find the value `0`, so this is the end and the result is `10`.*
+
+Provide an implementation for `sidewinder` so that it
+- follows the chain of indices to the end, and
+- returns the position of the end of the chain
+
+You may assume that `number_list` will never be an empty list, and that all of the values in the list are valid indices in the list.
+
```

Comment:
I've personally enjoyed and treasured providing the tests in `pytest` form whenever there's a code challenge. Is it possible for you to do that too?

Reply:
I started doing that for sidewinder, but then it really didn't make sense to do so for silly_sum since most of that is related to mocking the behavior of Input_int and just added a lot of noise. Since it didn't really work well for both, it seemed best to leave it off for both. That's why I added the message to the test error to give a sense of what was going on, while in sidewinder, we can see explicitly what list it's called with.

So I'd rather not add the explicit tests here.